internal/app/singularity/registry.go

```diff
@@ -134,6 +134,15 @@ func (l *Library) copyFromCache(hash, name, to string) error {
 		return errNotInCache
 	}
 
+	sylog.Debugf("Removing %s (if exists) before copying cached image", to)
+	err = os.Remove(to)
```

Comment:
I don't think this function is the right place for removal. Instead, you should remove image at a higher level, e.g. in `Pull`.

Reply:
Well, it was at a very high level before, (in the cli code). This was a issue because if a download fails, the overridden image will be lost. And that the point of this PR, to only remove the image at the very last moment. But I'm open to suggestions if you think otherwise...

Reply:
The other day I was working on something like this but I haven't pushed a PR yet because there's some very awkward logic around this.

My approach is to download to a temporary file in the same directory as the target. Check if the target exists _before_ downloading, if it's there check the --force flag and _not_ delete anything. Proceed with the download and check _again_ after the file has been downloaded, and if the --force flag was passed, nuke the existing file and rename the temporary file to the final name. Checking twice is necessary because the whole code around this is a little awkward, in the sense that the involved code does not take an io.Writer as parameter but instead everyone is dead set on getting the filename.

If the involved code took an io.Writer, things would not be simpler, you still have to download to a temporary location and check twice if the file is there, but at least the code would be a tiny bit less racy. It would make things simpler in the sense that you would not have to check for clobbering files all over the place.

So, I tend to agree with Sasha: the overwrite check should happen elsewhere. The reason why it's here is because this code contains some UI elements, as it will output warnings and what not. When I changed the code to work with a temporary file instead of the actual destination, I started to get messages on the console about the temporary filename, and that's actually the reason for not uploading my PR yet.

Reply:
Download to a temporary file in the same directory as the target, seems reasonable, I'll work on implementing that...

Reply:
And still you have this at a higher level, no need to remove here.

Reply:
So, "download to a temporary file in the same directory as the target." only seems useful if the cache is disabled
(`--disable-cache`). Unless I'm not understanding something? This seems like a good place to remove the final image (if overriding it), at the very last moment, just as the PR is suppose to fix.

In case there is some misunderstanding, heres the order-of-operation: (current behavior in this PR)

CLI:
 - check if the image already exists,
    - if force is false, abort
    - else if force is true, continue

PULL:
 - is cache disabled? if yes: pull to a temporary file in the same directory, then remove the 'to' image and rename the temp file to the 'to' image; else
    - check if the image is in cache (if not, download it)
    - remove the overwritten image (if exists)
    - copy the image from cache to the final destination